src/selectors_vm/ast.rs

```diff
@@ -1,86 +1,120 @@
 use super::parser::{Selector, SelectorImplDescriptor};
 use selectors::attr::{AttrSelectorOperator, ParsedCaseSensitivity};
 use selectors::parser::{Combinator, Component};
-use std::collections::HashSet;
-use std::fmt::Debug;
+use hashbrown::HashSet;
+use std::fmt::{self, Formatter, Debug};
 use std::hash::Hash;
 
-#[derive(Eq, PartialEq, Debug)]
-pub struct AttributeExprOperand {
-    pub name: String,
-    pub value: String,
-    pub case_sensitivity: ParsedCaseSensitivity,
+#[derive(PartialEq, Eq, Debug, Copy, Clone)]
+pub struct NthChild {
+    step: i32,
+    offset: i32,
 }
 
-#[derive(PartialEq, Eq, Debug)]
-pub enum AttributeExpr {
-    Id(String),
-    Class(String),
-    AttributeExists(String),
-    AttributeEqual(AttributeExprOperand),
-    AttributeIncludes(AttributeExprOperand),
-    AttributeDashMatch(AttributeExprOperand),
-    AttributePrefix(AttributeExprOperand),
-    AttributeSubstring(AttributeExprOperand),
-    AttributeSuffix(AttributeExprOperand),
+impl NthChild {
+    /// A first child with a step of 0 and an offset of 1
+    pub fn first() -> Self {
+        Self::new(0, 1)
+    }
+
+    pub fn new(step: i32, offset: i32) -> Self {
+        Self { step, offset }
+    }
+
+    pub fn has_index(self, index: i32) -> bool {
```

Comment:
It would be nice to have a unit test for this. This one of those arithmetic code parts that are easy to break.

Reply:
There is one added at the bottom (though I should add more case tests).